contracts/governance/governance.cpp

```diff
@@ -0,0 +1,650 @@
+      pevent.set_status( p.get_status() );
+      system::event( "proposal.status", pevent );
+
+      system::remove_object( state::proposal_space(), id );
+   }
+   else
+   {
+      p.set_status( governance::proposal_status::approved );
+      system::put_object( state::proposal_space(), id, p );
+
+      governance::proposal_status_event< system::detail::max_hash_size > pevent;
+      pevent.set_id( p.get_proposal().get_id() );
+      pevent.set_status( p.get_status() );
+      system::event( "proposal.status", pevent );
+   }
+}
+
+void handle_approved_proposal( proposal_record& p, uint64_t current_height )
+{
+   if ( p.get_vote_start_height() + constants::vote_period + constants::application_delay != current_height )
+      return;
+
+   std::string id( reinterpret_cast< const char* >( p.get_proposal().get_id().get_const() ), p.get_proposal().get_id().get_length() );
+
+   p.set_shall_authorize( true );
+   system::put_object( state::proposal_space(), id, p );
+
+   system::apply_transaction( p.get_proposal() );
+   p.set_status( governance::proposal_status::applied );
+
+   system::remove_object( state::proposal_space(), id );
+
+   governance::proposal_status_event< system::detail::max_hash_size > pevent;
+   pevent.set_id( p.get_proposal().get_id() );
+   pevent.set_status( p.get_status() );
+   system::event( "proposal.status", pevent );
+}
+
+void handle_votes()
+{
+   auto proposal_votes_message = system::get_block_field( "header.approved_proposals" ).message_value();
+   chain::list_type< 32, 64, system::detail::max_hash_size > proposal_votes;
+   proposal_votes_message.UnpackTo( proposal_votes );
+
+   for ( uint64_t i = 0 ; i < proposal_votes.get_values().get_length(); i++ )
+   {
+      auto proposal_vote = proposal_votes.get_values().get_const( i ).get_bytes_value();
+      std::string id( reinterpret_cast< const char* >( proposal_vote.get_const() ), proposal_vote.get_length() );
+
+      proposal_record prec;
+
+      if ( !system::get_object( state::proposal_space(), id, prec ) )
+         continue;
+
+      if ( prec.get_status() != governance::proposal_status::active )
+         continue;
+
+      auto current_vote_tally = prec.get_vote_tally();
+      if ( current_vote_tally != std::numeric_limits< uint64_t >::max() )
+         prec.set_vote_tally( current_vote_tally + 1 );
+
+      system::put_object( state::proposal_space(), id, prec );
+
+      governance::proposal_vote_event< system::detail::max_hash_size > pevent;
+      pevent.set_id( prec.get_proposal().get_id() );
+      pevent.set_vote_tally( prec.get_vote_tally() );
+      pevent.set_vote_threshold( prec.get_vote_threshold() );
+      system::event( "proposal.vote", pevent );
+   }
+}
+
+void block_callback()
+{
+   const auto [ caller, privilege ] = system::get_caller();
+   if ( privilege != chain::privilege::kernel_mode )
+   {
+      system::log( "Governance contract block callback must be called from kernel" );
+      system::exit_contract( 1 );
+   }
+
```

Comment:
I propose to call `system::require_authority` here. I see 3 advantages:
1. The mining pool can be totally governed by a DAO, which could determine who can mine in its name and who doesn't.
2. The DAO can validate that the proposal votes in the headers are the correct ones. Then the DAO can control the votes as well.
3. The thunk `process_block_signature` can be removed since `pre_block_callback` has similar functionalities.